cmd/bpf2go/gen/output.go

```diff
@@ -141,22 +140,50 @@ func Generate(args GenerateArgs) error {
 		programs[name] = args.Identifier(name)
 	}
 
-	typeNames := make(map[btf.Type]string)
-	for _, typ := range args.Types {
-		// NB: This also deduplicates types.
-		typeNames[typ] = args.Stem + args.Identifier(typ.TypeName())
+	tn := templateName(args.Stem)
+	reservedNames := map[string]struct{}{
+		tn.Specs():         {},
+		tn.MapSpecs():      {},
+		tn.ProgramSpecs():  {},
+		tn.VariableSpecs(): {},
+		tn.Objects():       {},
+		tn.Maps():          {},
+		tn.Programs():      {},
+		tn.Variables():     {},
 	}
 
-	// Ensure we don't have conflicting names and generate a sorted list of
-	// named types so that the output is stable.
-	types, err := sortTypes(typeNames)
-	if err != nil {
-		return err
+	typeByName := map[string]btf.Type{}
+	nameByType := map[btf.Type]string{}
+	for _, typ := range args.Types {
+		// NB: This also deduplicates types.
+		name := args.Stem + args.Identifier(typ.TypeName())
+		if _, reserved := reservedNames[name]; reserved {
+			return fmt.Errorf("type name %q is reserved", name)
+		}
+		if otherType, ok := typeByName[name]; ok {
+			if otherType == typ {
+				continue
+			}
+			return fmt.Errorf("type name %q is used multiple times", name)
+		}
+		typeByName[name] = typ
+		nameByType[typ] = name
 	}
 
 	gf := &btf.GoFormatter{
-		Names:      typeNames,
+		Names:      nameByType,
 		Identifier: args.Identifier,
+		ShortEnumIdentifier: func(_, element string) string {
```

Comment:
Could this be made a part of `EnumIdentifier` instead? Not sure why this deserves its own hook, it's just another step during enum identifier generation.

Reply:
>  Not sure why this deserves its own hook, it's just another step during enum identifier generation.

I propose generating BOTH long and short identifiers to maintain backwards compatibility.